[Bug]: Tests fail with Go1.20

### Steps to Reproduce

With the upcoming Go1.20, the tests fail.



### Your Environment

* OS - Linux
* Version - `go version go1.20rc2 linux/amd64`

### Expected Behavior

Tests pass

### Actual Behavior

```
--- FAIL: TestGenerateDefaultKeyPair (0.00s)
    --- FAIL: TestGenerateDefaultKeyPair/ok (0.00s)
        key_test.go:161: GenerateDefaultKeyPair() error = error generating EC key: unexpected EOF, wantErr false
--- FAIL: TestGenerateKey (0.04s)
    --- FAIL: TestGenerateKey/P-256 (0.00s)
        key_test.go:221: GenerateKey() = &{{0x74b810 35852464506455460549044114594740426399053290184916918376269928775285032380340 35592537140215175340666390858275148912552451716305781313406699830287187232976} 67950854095978885778709149645661493900291163920680215092956399472860327293574}, want &{{0x74b810 19704518974308737877293353907502558429045798270165099103081372071168779902409 58896009669746436177740100186299804454165317925669887304774016084906450088580} 11120941389810810843532862460367856571792073421468959968906009436182716557234}
--- FAIL: TestGenerateKeyPair (0.01s)
    --- FAIL: TestGenerateKeyPair/P-384 (0.00s)
        key_test.go:311: GenerateKeyPair() got = &{0x74b820 24697468251312879413921642862667409471842541489078317592121694595686428419475113887987400634869787941119330228571259 13698494916488625805005405466483475817819606390473491312445141184589955251846558222319225805149063393341564192937842}, want &{0x74b820 35454505642517037596442009845453909192969625637970734507958617492367098887700066346449716986504713615026807959376657 6891759594063328547705138422554547542138416624972531425875556538308946341781185598123836221693510700765083687681201}
        key_test.go:314: GenerateKeyPair() got1 = &{{0x74b820 24697468251312879413921642862667409471842541489078317592121694595686428419475113887987400634869787941119330228571259 13698494916488625805005405466483475817819606390473491312445141184589955251846558222319225805149063393341564192937842} 2957881366700156430588424677089287375999430865732298546418503328650718709931263488104620184716073974754309510975324}, want &{{0x74b820 35454505642517037596442009845453909192969625637970734507958617492367098887700066346449716986504713615026807959376657 6891759594063328547705138422554547542138416624972531425875556538308946341781185598123836221693510700765083687681201} 8579512143744941197334955432128905959339255488624941192119277453915276083956451337911708444501540361777302993853566}
    --- FAIL: TestGenerateKeyPair/P-521 (0.00s)
        key_test.go:311: GenerateKeyPair() got = &{0x74b830 3047592644937981874635237818065674803617375273409961699714870541836272696003181581093651283242621442829116604166880022290888069713990826092967328231043774188 789262853125457302002107880500230696678257320415205189566310196214326981614549566489789978195853678138406198502965693640973229467040712549378511967363684946}, want &{0x74b830 6113081197798226192372987184858317167839636419336792855211995915536211147897478046796557365674421374133784995766352826587480679667182314660273705278401927851 637570240195211632955561724572269277585015284066907480669942629149935255078091820215825193728570369035135214388970989361694077286977274692961077336052556388}
        key_test.go:314: GenerateKeyPair() got1 = &{{0x74b830 3047592644937981874635237818065674803617375273409961699714870541836272696003181581093651283242621442829116604166880022290888069713990826092967328231043774188 789262853125457302002107880500230696678257320415205189566310196214326981614549566489789978195853678138406198502965693640973229467040712549378511967363684946} 847584683012100394816627314446588107582753362451945123532648688414257577382561050235550624451002656822367225352986730985221226023034049527358083080620739824}, want &{{0x74b830 6113081197798226192372987184858317167839636419336792855211995915536211147897478046796557365674421374133784995766352826587480679667182314660273705278401927851 637570240195211632955561724572269277585015284066907480669942629149935255078091820215825193728570369035135214388970989361694077286977274692961077336052556388} 740552556983495051126717327262669197197648833183831228432465269705329532909463218440331178620861664741175881532956104047291239905642398907587377468785094824}
    --- FAIL: TestGenerateKeyPair/Ed25519 (0.00s)
        key_test.go:311: GenerateKeyPair() got = [13 43 182 253 164 206 147 175 84 56 238 53 240 152 47 89 194 146 190 88 38 99 94 229 128 182 162 19 131 151 150 32], want [24 116 117 218 248 82 198 4 191 170 83 77 232 132 76 0 210 62 135 5 57 208 178 151 168 92 29 201 97 117 118 161]
        key_test.go:314: GenerateKeyPair() got1 = [168 207 70 239 199 219 209 61 14 59 165 226 35 140 207 204 77 154 10 216 252 220 219 160 139 158 237 12 47 133 158 193 13 43 182 253 164 206 147 175 84 56 238 53 240 152 47 89 194 146 190 88 38 99 94 229 128 182 162 19 131 151 150 32], want [207 70 239 199 219 209 61 14 59 165 226 35 140 207 204 77 154 10 216 252 220 219 160 139 158 237 12 47 133 158 193 188 24 116 117 218 248 82 198 4 191 170 83 77 232 132 76 0 210 62 135 5 57 208 178 151 168 92 29 201 97 117 118 161]
    --- FAIL: TestGenerateKeyPair/eof (0.00s)
        key_test.go:307: GenerateKeyPair() error = <nil>, wantErr true
--- FAIL: TestGenerateSigner (0.01s)
    --- FAIL: TestGenerateSigner/P-256 (0.00s)
        key_test.go:435: GenerateSigner() = &{{0x74b810 73364206300223275984623165552918181084247314448355077616775976124826204201962 56284526082827190149484766087292813359785585803052020954966551130509863245629} 16399757394950745931281014142200954873384491614817309221586503933123965988998}, want &{{0x74b810 22248328256213248797325976063267447311511351580655081495882005070640566121552 91042212017445225092862545181703136754516504537320715679784975537765216651940} 29822680564007923159384160090679764868710405430170855305671982594862626146013}
    --- FAIL: TestGenerateSigner/eof (0.00s)
        key_test.go:431: GenerateSigner() error = <nil>, wantErr true
FAIL
FAIL    go.step.sm/crypto/keyutil       8.393s
ok      go.step.sm/crypto/kms   0.053s
ok      go.step.sm/crypto/kms/apiv1     0.007s
ok      go.step.sm/crypto/kms/awskms    0.019s
?       go.step.sm/crypto/kms/azurekms/internal/mock    [no test files]
go build github.com/go-piv/piv-go/piv:
# pkg-config --cflags  -- libpcsclite
pkg-config: exec: "pkg-config": executable file not found in $PATH
ok      go.step.sm/crypto/kms/azurekms  5.816s
ok      go.step.sm/crypto/kms/cloudkms  10.015s
ok      go.step.sm/crypto/kms/pkcs11    31.162s
ok      go.step.sm/crypto/kms/softkms   1.679s
ok      go.step.sm/crypto/kms/sshagentkms       0.065s
ok      go.step.sm/crypto/kms/uri       0.004s
FAIL    go.step.sm/crypto/kms/yubikey [build failed]
ok      go.step.sm/crypto/minica        0.042s
ok      go.step.sm/crypto/pemutil       18.017s
ok      go.step.sm/crypto/randutil      0.005s
ok      go.step.sm/crypto/sshutil       1.820s
ok      go.step.sm/crypto/tlsutil       4.141s
ok      go.step.sm/crypto/x25519        0.573s
ok      go.step.sm/crypto/x509util      2.763s
FAIL
```

### Additional Context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction.
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/jose/generate_test.go
+++ b/jose/generate_test.go
@@ -9,6 +9,7 @@ import (
 	"crypto/x509"
 	"encoding/json"
 	"encoding/pem"
+	"io"
 	"math/big"
 	"os"
 	"reflect"
@@ -367,12 +368,32 @@ func tempFile(t *testing.T) (*os.File, func()) {
 	}
 }
 
+type mockReader struct{}
+
+func (mockReader) Read(buf []byte) (int, error) {
+	for i := range buf {
+		buf[i] = byte(i % 256)
+	}
+	return len(buf), nil
+}
+
+type eofReader struct{}
+
+func (eofReader) Read(buf []byte) (int, error) {
+	return 0, io.EOF
+}
+
 func TestGenerateDefaultKeyPair(t *testing.T) {
-	reader := mustTeeReader(t)
+	rr := rand.Reader
+	t.Cleanup(func() {
+		rand.Reader = rr
+		jose.RandReader = rr
+	})
+
+	rand.Reader = mockReader{}
+	jose.RandReader = mockReader{}
 	jwk := mustGenerateJWK(t, "EC", "P-256", "ES256", "sig", "", 0)
 	jwe := mustEncryptJWK(t, jwk, []byte("planned password"))
-	rand.Reader = reader
-	jose.RandReader = reader
 
 	var err error
 	if jwk.KeyID, err = Thumbprint(jwk); err != nil {
@@ -382,6 +403,7 @@ func TestGenerateDefaultKeyPair(t *testing.T) {
 
 	type args struct {
 		passphrase []byte
+		randReader io.Reader
 	}
 	tests := []struct {
 		name           string
@@ -391,13 +413,15 @@ func TestGenerateDefaultKeyPair(t *testing.T) {
 		want1Decrypted *JSONWebKey
 		wantErr        bool
 	}{
-		{"ok", args{[]byte("planned password")}, &jwkPub, jwe, jwk, false},
-		{"failEmptyPassword", args{[]byte("")}, nil, nil, nil, true},
-		{"failNilPassword", args{nil}, nil, nil, nil, true},
-		{"failEOF", args{[]byte("planned password")}, nil, nil, nil, true},
+		{"ok", args{[]byte("planned password"), mockReader{}}, &jwkPub, jwe, jwk, false},
+		{"failEmptyPassword", args{[]byte(""), rr}, nil, nil, nil, true},
+		{"failNilPassword", args{nil, rr}, nil, nil, nil, true},
+		{"failEOF", args{[]byte("planned password"), eofReader{}}, nil, nil, nil, true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.args.randReader
+			jose.RandReader = tt.args.randReader
 			got, got1, err := GenerateDefaultKeyPair(tt.args.passphrase)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateDefaultKeyPair() error = %v, wantErr %v", err, tt.wantErr)

--- a/keyutil/key_test.go
+++ b/keyutil/key_test.go
@@ -99,7 +99,7 @@ func verifyKeyPair(h crypto.Hash, priv, pub interface{}) error {
 	}
 	sig, err := s.Sign(randReader, sum, h)
 	if err != nil {
-		return fmt.Errorf("%T.Sign() error = %v", s, err)
+		return fmt.Errorf("%T.Sign() error = %w", s, err)
 	}
 
 	switch p := pub.(type) {

--- a/keyutil/key_test.go
+++ b/keyutil/key_test.go
@@ -499,7 +499,10 @@ func TestGenerateDefaultSigner(t *testing.T) {
 func TestGenerateSigner(t *testing.T) {
 	assertSigner := func(h crypto.Hash) func(t *testing.T, got crypto.Signer) {
 		return func(t *testing.T, got crypto.Signer) {
-
+			t.Helper()
+			if err := verifyPrivateKey(h, got); err != nil {
+				t.Errorf("GenerateSigner() error = %v", err)
+			}
 		}
 	}
 	assertNil := func() func(t *testing.T, got crypto.Signer) {

--- a/keyutil/key_test.go
+++ b/keyutil/key_test.go
@@ -1,7 +1,6 @@
 package keyutil
 
 import (
-	"bytes"
 	"crypto"
 	"crypto/ecdsa"
 	"crypto/ed25519"
@@ -9,6 +8,7 @@ import (
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/pem"
+	"fmt"
 	"io"
 	"reflect"
 	"testing"
@@ -61,13 +61,74 @@ func must(args ...interface{}) interface{} {
 	return args[0]
 }
 
-func setTeeReader(t *testing.T, w *bytes.Buffer) {
-	t.Helper()
-	reader := rand.Reader
+var randReader = rand.Reader
+
+func cleanupRandReader(t *testing.T) {
+	rr := rand.Reader
 	t.Cleanup(func() {
-		rand.Reader = reader
+		rand.Reader = rr
 	})
-	rand.Reader = io.TeeReader(reader, w)
+}
+
+type zeroReader struct{}
+
+func (zeroReader) Read(buf []byte) (int, error) {
+	for i := range buf {
+		buf[i] = 0
+	}
+	return len(buf), nil
+}
+
+type eofReader struct{}
+
+func (eofReader) Read(buf []byte) (int, error) {
+	return 0, io.EOF
+}
+
+func verifyKeyPair(h crypto.Hash, priv, pub interface{}) error {
+	s, ok := priv.(crypto.Signer)
+	if !ok {
+		return fmt.Errorf("type %T is not a crypto.Signer", priv)
+	}
+
+	var sum []byte
+	if h == crypto.Hash(0) {
+		sum = []byte("a message")
+	} else {
+		sum = h.New().Sum([]byte("a message"))
+	}
+	sig, err := s.Sign(randReader, sum, h)
+	if err != nil {
+		return fmt.Errorf("%T.Sign() error = %v", s, err)
+	}
+
+	switch p := pub.(type) {
+	case *ecdsa.PublicKey:
+		if !ecdsa.VerifyASN1(p, sum, sig) {
+			return fmt.Errorf("ecdsa.VerifyASN1 failed")
+		}
+	case *rsa.PublicKey:
+		if err := rsa.VerifyPKCS1v15(p, h, sig, sum); err != nil {
+			return fmt.Errorf("rsa.VerifyPKCS1v15 failed")
+		}
+	case ed25519.PublicKey:
+		if !ed25519.Verify(p, sum, sig) {
+			return fmt.Errorf("ed25519.Verify failed")
+		}
+	default:
+		return fmt.Errorf("unsupported public key type %T", pub)
+	}
+
+	return nil
+}
+
+func verifyPrivateKey(h crypto.Hash, priv interface{}) error {
+	s, ok := priv.(crypto.Signer)
+	if !ok {
+		return fmt.Errorf("type %T is not a crypto.Signer", priv)
+	}
+
+	return verifyKeyPair(h, priv, s.Public())
 }
 
 func TestPublicKey(t *testing.T) {
@@ -111,80 +172,105 @@ func TestPublicKey(t *testing.T) {
 }
 
 func TestGenerateDefaultKey(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	ecdsaKey, err := generateECKey("P-256")
-	assert.FatalError(t, err)
-	rand.Reader = buf
+	cleanupRandReader(t)
 
 	tests := []struct {
-		name    string
-		want    interface{}
-		wantErr bool
+		name      string
+		rr        io.Reader
+		assertion func(t *testing.T, got interface{})
+		wantErr   bool
 	}{
-		{"ok", ecdsaKey, false},
-		{"eof", nil, true},
+		{"ok", randReader, func(t *testing.T, got interface{}) {
+			t.Helper()
+			if err := verifyPrivateKey(crypto.SHA256, got); err != nil {
+				t.Errorf("GenerateDefaultKey() error = %v", err)
+			}
+		}, false},
+		{"eof", eofReader{}, func(t *testing.T, got interface{}) {
+			if !reflect.DeepEqual(got, nil) {
+				t.Errorf("GenerateDefaultKey() got = %v, want nil", got)
+			}
+		}, true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.rr
 			got, err := GenerateDefaultKey()
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateDefaultKey() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateDefaultKey() = %T, want %v", got, tt.want)
-			}
+			tt.assertion(t, got)
 		})
 	}
 }
 
 func TestGenerateDefaultKeyPair(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	ecdsaKey := must(generateECKey("P-256")).(*ecdsa.PrivateKey)
-	rand.Reader = buf
+	cleanupRandReader(t)
+
+	assertKey := func(h crypto.Hash) func(t *testing.T, got, got1 interface{}) {
+		return func(t *testing.T, got, got1 interface{}) {
+			t.Helper()
+			if err := verifyKeyPair(h, got1, got); err != nil {
+				t.Errorf("GenerateDefaultKeyPair() error = %v", err)
+			}
+		}
+	}
+
+	assertNil := func() func(t *testing.T, got, got1 interface{}) {
+		return func(t *testing.T, got, got1 interface{}) {
+			t.Helper()
+			if !reflect.DeepEqual(got, nil) {
+				t.Errorf("GenerateDefaultKeyPair() got = %v, want nil", got)
+			}
+			if !reflect.DeepEqual(got1, nil) {
+				t.Errorf("GenerateDefaultKeyPair() got1 = %v, want nil", got1)
+			}
+		}
+	}
 
 	tests := []struct {
-		name    string
-		want    interface{}
-		want1   interface{}
-		wantErr bool
+		name      string
+		rr        io.Reader
+		assertion func(t *testing.T, got, got1 interface{})
+		wantErr   bool
 	}{
-		{"ok", ecdsaKey.Public(), ecdsaKey, false},
-		{"eof", nil, nil, true},
+		{"ok", randReader, assertKey(crypto.SHA256), false},
+		{"eof", eofReader{}, assertNil(), true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.rr
 			got, got1, err := GenerateDefaultKeyPair()
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateDefaultKeyPair() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateDefaultKeyPair() got = %v, want %v", got, tt.want)
-			}
-			if !reflect.DeepEqual(got1, tt.want1) {
-				t.Errorf("GenerateDefaultKeyPair() got1 = %v, want %v", got1, tt.want1)
-			}
+			tt.assertion(t, got, got1)
 		})
 	}
 }
 
 func TestGenerateKey(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	p256Key, err := generateECKey("P-256")
-	assert.FatalError(t, err)
-	p384Key, err := generateECKey("P-384")
-	assert.FatalError(t, err)
-	p521Key, err := generateECKey("P-521")
-	assert.FatalError(t, err)
-	ed25519Key, err := generateOKPKey("Ed25519")
-	assert.FatalError(t, err)
-	octKey, err := generateOctKey(32)
-	assert.FatalError(t, err)
-	rand.Reader = buf
+	cleanupRandReader(t)
+
+	assertKey := func(t *testing.T, h crypto.Hash, key interface{}) {
+		t.Helper()
+		if err := verifyPrivateKey(h, key); err != nil {
+			t.Errorf("GenerateKey() error = %v", err)
+		}
+	}
+
+	octKey := make([]byte, 32)
+	for i := range octKey {
+		octKey[i] = 'a'
+	}
+	assertOCT := func(t *testing.T, h crypto.Hash, key interface{}) {
+		t.Helper()
+		if !reflect.DeepEqual(key, octKey) {
+			t.Errorf("GenerateKey() got = %v, want %v", key, octKey)
+		}
+	}
 
 	type args struct {
 		kty  string
@@ -193,32 +279,35 @@ func TestGenerateKey(t *testing.T) {
 	}
 	tests := []struct {
 		name    string
+		rr      io.Reader
 		args    args
-		want    crypto.PrivateKey
+		assert  func(t *testing.T, h crypto.Hash, key interface{})
+		hash    crypto.Hash
 		wantErr bool
 	}{
-		{"P-256", args{"EC", "P-256", 0}, p256Key, false},
-		{"P-384", args{"EC", "P-384", 0}, p384Key, false},
-		{"P-521", args{"EC", "P-521", 0}, p521Key, false},
-		{"Ed25519", args{"OKP", "Ed25519", 0}, ed25519Key, false},
-		{"OCT", args{"oct", "", 32}, octKey, false},
-		{"eof EC", args{"EC", "P-256", 0}, nil, true},
-		{"eof RSA", args{"RSA", "", 1024}, nil, true},
-		{"eof OKP", args{"OKP", "Ed25519", 0}, nil, true},
-		{"eof oct", args{"oct", "", 32}, nil, true},
-		{"unknown EC curve", args{"EC", "P-128", 0}, nil, true},
-		{"unknown OKP curve", args{"OKP", "Edward", 0}, nil, true},
-		{"unknown type", args{"FOO", "", 1024}, nil, true},
+		{"P-256", randReader, args{"EC", "P-256", 0}, assertKey, crypto.SHA256, false},
+		{"P-384", randReader, args{"EC", "P-384", 0}, assertKey, crypto.SHA384, false},
+		{"P-521", randReader, args{"EC", "P-521", 0}, assertKey, crypto.SHA512, false},
+		{"Ed25519", randReader, args{"OKP", "Ed25519", 0}, assertKey, crypto.Hash(0), false},
+		{"OCT", zeroReader{}, args{"oct", "", 32}, assertOCT, crypto.Hash(0), false},
+		{"eof EC", eofReader{}, args{"EC", "P-256", 0}, nil, 0, true},
+		{"eof RSA", eofReader{}, args{"RSA", "", 1024}, nil, 0, true},
+		{"eof OKP", eofReader{}, args{"OKP", "Ed25519", 0}, nil, 0, true},
+		{"eof oct", eofReader{}, args{"oct", "", 32}, nil, 0, true},
+		{"unknown EC curve", randReader, args{"EC", "P-128", 0}, nil, 0, true},
+		{"unknown OKP curve", randReader, args{"OKP", "Edward", 0}, nil, 0, true},
+		{"unknown type", randReader, args{"FOO", "", 1024}, nil, 0, true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.rr
 			got, err := GenerateKey(tt.args.kty, tt.args.crv, tt.args.size)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateKey() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateKey() = %v, want %v", got, tt.want)
+			if !tt.wantErr {
+				tt.assert(t, tt.hash, got)
 			}
 		})
 	}
@@ -269,15 +358,28 @@ func TestGenerateKey_rsa(t *testing.T) {
 }
 
 func TestGenerateKeyPair(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	p256Key := must(generateECKey("P-256")).(*ecdsa.PrivateKey)
-	p384Key := must(generateECKey("P-384")).(*ecdsa.PrivateKey)
-	p521Key := must(generateECKey("P-521")).(*ecdsa.PrivateKey)
-	ed25519Key := must(generateOKPKey("Ed25519")).(ed25519.PrivateKey)
-	_, err := generateOctKey(32)
-	assert.FatalError(t, err)
-	rand.Reader = buf
+	cleanupRandReader(t)
+
+	assertKey := func(h crypto.Hash) func(t *testing.T, got, got1 interface{}) {
+		return func(t *testing.T, got, got1 interface{}) {
+			t.Helper()
+			if err := verifyKeyPair(h, got1, got); err != nil {
+				t.Errorf("GenerateKeyPair() error = %v", err)
+			}
+		}
+	}
+
+	assertNil := func() func(t *testing.T, got, got1 interface{}) {
+		return func(t *testing.T, got, got1 interface{}) {
+			t.Helper()
+			if !reflect.DeepEqual(got, nil) {
+				t.Errorf("GenerateKeyPair() got = %v, want nil", got)
+			}
+			if !reflect.DeepEqual(got1, nil) {
+				t.Errorf("GenerateKeyPair() got1 = %v, want nil", got1)
+			}
+		}
+	}
 
 	type args struct {
 		kty  string
@@ -285,34 +387,30 @@ func TestGenerateKeyPair(t *testing.T) {
 		size int
 	}
 	tests := []struct {
-		name    string
-		args    args
-		want    crypto.PublicKey
-		want1   crypto.PrivateKey
-		wantErr bool
+		name      string
+		rr        io.Reader
+		args      args
+		assertion func(t *testing.T, got, got1 interface{})
+		wantErr   bool
 	}{
-		{"P-256", args{"EC", "P-256", 0}, p256Key.Public(), p256Key, false},
-		{"P-384", args{"EC", "P-384", 0}, p384Key.Public(), p384Key, false},
-		{"P-521", args{"EC", "P-521", 0}, p521Key.Public(), p521Key, false},
-		{"Ed25519", args{"OKP", "Ed25519", 0}, ed25519Key.Public(), ed25519Key, false},
-		{"OCT", args{"oct", "", 32}, nil, nil, true},
-		{"eof", args{"EC", "P-256", 0}, nil, nil, true},
-		{"unknown", args{"EC", "P-128", 0}, nil, nil, true},
-		{"unknown", args{"FOO", "", 1024}, nil, nil, true},
+		{"P-256", randReader, args{"EC", "P-256", 0}, assertKey(crypto.SHA256), false},
+		{"P-384", randReader, args{"EC", "P-384", 0}, assertKey(crypto.SHA384), false},
+		{"P-521", randReader, args{"EC", "P-521", 0}, assertKey(crypto.SHA512), false},
+		{"Ed25519", randReader, args{"OKP", "Ed25519", 0}, assertKey(crypto.Hash(0)), false},
+		{"OCT", zeroReader{}, args{"oct", "", 32}, assertNil(), true},
+		{"eof", eofReader{}, args{"EC", "P-256", 0}, assertNil(), true},
+		{"unknown", randReader, args{"EC", "P-128", 0}, assertNil(), true},
+		{"unknown", randReader, args{"FOO", "", 1024}, assertNil(), true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.rr
 			got, got1, err := GenerateKeyPair(tt.args.kty, tt.args.crv, tt.args.size)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateKeyPair() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateKeyPair() got = %v, want %v", got, tt.want)
-			}
-			if !reflect.DeepEqual(got1, tt.want1) {
-				t.Errorf("GenerateKeyPair() got1 = %v, want %v", got1, tt.want1)
-			}
+			tt.assertion(t, got, got1)
 		})
 	}
 }
@@ -365,44 +463,53 @@ func TestGenerateKeyPair_rsa(t *testing.T) {
 }
 
 func TestGenerateDefaultSigner(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	ecdsaKey, err := generateECKey("P-256")
-	assert.FatalError(t, err)
-	rand.Reader = buf
+	cleanupRandReader(t)
 
 	tests := []struct {
-		name    string
-		want    interface{}
-		wantErr bool
+		name      string
+		rr        io.Reader
+		assertion func(t *testing.T, got crypto.Signer)
+		wantErr   bool
 	}{
-		{"ok", ecdsaKey, false},
-		{"eof", nil, true},
+		{"ok", randReader, func(t *testing.T, got crypto.Signer) {
+			t.Helper()
+			if err := verifyPrivateKey(crypto.SHA256, got); err != nil {
+				t.Errorf("GenerateDefaultSigner() error = %v", err)
+			}
+		}, false},
+		{"eof", eofReader{}, func(t *testing.T, got crypto.Signer) {
+			if !reflect.DeepEqual(got, nil) {
+				t.Errorf("GenerateDefaultSigner() got = %v, want nil", got)
+			}
+		}, true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			rand.Reader = tt.rr
 			got, err := GenerateDefaultSigner()
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GenerateDefaultSigner() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateDefaultSigner() = %T, want %v", got, tt.want)
-			}
+			tt.assertion(t, got)
 		})
 	}
 }
 
 func TestGenerateSigner(t *testing.T) {
-	buf := new(bytes.Buffer)
-	setTeeReader(t, buf)
-	p256Key := must(generateECKey("P-256")).(*ecdsa.PrivateKey)
-	p384Key := must(generateECKey("P-384")).(*ecdsa.PrivateKey)
-	p521Key := must(generateECKey("P-521")).(*ecdsa.PrivateKey)
-	ed25519Key := must(generateOKPKey("Ed25519")).(ed25519.PrivateKey)
-	_, err := generateOctKey(32)
-	assert.FatalError(t, err)
-	rand.Reader = buf
+	assertSigner := func(h crypto.Hash) func(t *testing.T, got crypto.Signer) {
+		return func(t *testing.T, got crypto.Signer) {
+
+		}
+	}
+	assertNil := func() func(t *testing.T, got crypto.Signer) {
+		return func(t *testing.T, got crypto.Signer) {
+			t.Helper()
+			if !reflect.DeepEqual(got, nil) {
+				t.Errorf("GenerateSigner() got = %v, want nil", got)
+			}
+		}
+	}
 
 	type args struct {
 		kty  string
@@ -410,19 +517,18 @@ func TestGenerateSigner(t *testing.T) {
 		size int
 	}
 	tests := []struct {
-		name    string
-		args    args
-		want    crypto.Signer
-		wantErr bool
+		name      string
+		args      args
+		assertion func(t *testing.T, got crypto.Signer)
+		wantErr   bool
 	}{
-		{"P-256", args{"EC", "P-256", 0}, p256Key, false},
-		{"P-384", args{"EC", "P-384", 0}, p384Key, false},
-		{"P-521", args{"EC", "P-521", 0}, p521Key, false},
-		{"Ed25519", args{"OKP", "Ed25519", 0}, ed25519Key, false},
-		{"OCT", args{"oct", "", 32}, nil, true},
-		{"eof", args{"EC", "P-256", 0}, nil, true},
-		{"unknown", args{"EC", "P-128", 0}, nil, true},
-		{"unknown", args{"FOO", "", 1024}, nil, true},
+		{"P-256", args{"EC", "P-256", 0}, assertSigner(crypto.SHA256), false},
+		{"P-384", args{"EC", "P-384", 0}, assertSigner(crypto.SHA384), false},
+		{"P-521", args{"EC", "P-521", 0}, assertSigner(crypto.SHA512), false},
+		{"Ed25519", args{"OKP", "Ed25519", 0}, assertSigner(crypto.Hash(0)), false},
+		{"OCT", args{"oct", "", 32}, assertNil(), true},
+		{"unknown", args{"EC", "P-128", 0}, assertNil(), true},
+		{"unknown", args{"FOO", "", 1024}, assertNil(), true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
@@ -431,9 +537,7 @@ func TestGenerateSigner(t *testing.T) {
 				t.Errorf("GenerateSigner() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("GenerateSigner() = %v, want %v", got, tt.want)
-			}
+			tt.assertion(t, got)
 		})
 	}
 }

